extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -0,0 +1,735 @@
+    addDeviceButton.classList.add('i-amphtml-story-dev-tools-button');
+    addDeviceButton.setAttribute(
+      'data-action',
+      PREVIEW_ACTIONS.SHOW_ADD_DEVICE_DIALIG
+    );
+    addDeviceButton.querySelector(
+      '.i-amphtml-story-dev-tools-device-chip-name'
+    ).textContent = 'ADD DEVICE';
+    return addDeviceButton;
+  }
+
+  /**
+   * Builds the add device button
+   * @private
+   * @return {!Element}
+   */
+  buildHelpButton_() {
+    const addHelpButton = buildHelpButtonTemplate(this.element);
+    addHelpButton.setAttribute('data-action', PREVIEW_ACTIONS.SHOW_HELP_DIALOG);
+    return addHelpButton;
+  }
+
+  /**
+   * Creates a device layout for preview
+   * @private
+   * @param {!DeviceInfo} device
+   * @return {!Element}
+   */
+  buildDeviceLayout_(device) {
+    const deviceLayout = buildDeviceTemplate(this.element);
+    const devicePlayer = deviceLayout.querySelector('amp-story-player');
+    devicePlayer.setAttribute('width', device.width);
+    devicePlayer.setAttribute('height', device.height);
+    const storyA = devicePlayer.querySelector('a');
+    storyA.textContent = 'Story 1';
+    storyA.href = this.storyUrl_;
+    setStyles(devicePlayer, {
+      width: device.width + 'px',
+      height: device.height + 'px',
+    });
+    const playerImpl = new AmpStoryPlayer(this.win, devicePlayer);
+    playerImpl.load();
+    setStyles(
+      deviceLayout.querySelector('.i-amphtml-story-dev-tools-device-screen'),
+      {
+        height: device.deviceHeight
+          ? device.deviceHeight + 'px'
+          : 'fit-content',
+      }
+    );
+    device.player = devicePlayer;
+    return deviceLayout;
+  }
+
+  /**
+   * @private
+   * @param {!Element} targetElement
+   */
+  handleTap_(targetElement) {
+    const actionElement = closest(
+      targetElement,
+      (el) => el.hasAttribute('data-action'),
+      this.element
+    );
+    if (actionElement) {
+      switch (actionElement.getAttribute('data-action')) {
+        case PREVIEW_ACTIONS.SHOW_HELP_DIALOG:
+          this.showHelpDialog_();
+          break;
+        case PREVIEW_ACTIONS.SHOW_ADD_DEVICE_DIALIG:
+          this.showAddDeviceDialog_();
+          break;
+        case PREVIEW_ACTIONS.CLOSE_DIALOG:
+          this.hideCurrentDialog_();
+          break;
+        case PREVIEW_ACTIONS.REMOVE_DEVICE:
+        case PREVIEW_ACTIONS.TOGGLE_DEVICE_CHIP:
+          this.onDeviceChipToggled_(actionElement);
+          break;
+      }
+    }
+  }
+
+  /**
+   * @private
+   * @param {string} deviceName
+   */
+  addDevice_(deviceName) {
+    const deviceSpecs = ALL_DEVICES.find((d) => d.name === deviceName);
+    if (!deviceSpecs) {
+      return;
+    }
+    const deviceLayout = this.buildDeviceLayout_(deviceSpecs, this.storyUrl_);
+    deviceSpecs.element = deviceLayout;
+    deviceSpecs.chip = this.buildDeviceChip_(deviceSpecs.name);
+    this.mutateElement(() => {
+      this.element.appendChild(deviceLayout);
```

Comment:
Putting these in a wrapper could make the dom tree more logical.
You could then put `perspective: 100px;` on the wrapper instead of inlining it on each element.

Reply:
No benefits or drawbacks really, besides saving a `queryElement`call. Will add a devices container

Reply:
Tried for a while to make the `perspective: 100px` on the parent work, but it changes the frame of reference for the transform operations so the math to position all the items become **much** harder to the point where it's better to set the perspective on each device better, so they are scaled with the origin as the center of the device instead of the center of the container.

TL;DR: Having the perspective origin on the center of each device is better/easier than having it on the center of the container.